File: problemas/pilha.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>

using namespace std;

int main()
{
    int n;
    while(1)
    {
        queue<int> cards;
        cin >> n;
        if (n == 0) return 0;
        cout << "Discarded cards: ";
        for (int i = 0; i < n; i++)
        {
            cards.push(i + 1);
        }

        while (cards.size() > 1)
        {
            cout << cards.front();
            if (cards.size() > 2) cout << ", ";
            cards.pop();
            cards.push(cards.front());
            cards.pop();
        }
        cout << "\nRemaining card: " << cards.front() << endl;
    }
}

File: tutoria-samuel/pilha.py
t = int(input())

while t > 0:
	t -= 1
	stack = [];
	s = input()
	ans = 1
	for c in s:
		if c == '{' or c == '[' or c == '(':
			stack.append(c);
		elif c == ']':
			if len(stack) == 0 or stack[-1] != '[':
				ans = 0
			else:
				stack.pop()
		elif c == ')':
			if len(stack) == 0 or stack[-1] != '(':
				ans = 0
			else:
				stack.pop()
		elif c == '}':
			if len(stack) == 0 or stack[-1] != '{':
				ans = 0
			else:
				stack.pop()
		
	if ans and len(stack) == 0:
		print("S")
	else:
		print("N")



	
